File: frontend/src/api/neighborhoods.ts
import { Neighborhood, FeaturedNeighborhoodsResponse } from "../types/api";
import { fetchApi } from "./client";

export interface NeighborhoodQuery {
  sort_by?: "median_sold_price" | "average_sold_price" | "avg_asking_price" | "locality";
  limit?: number;
}

export async function getNeighborhoods(query: NeighborhoodQuery = {}): Promise<Neighborhood[]> {
  const params = new URLSearchParams();
  if (query.sort_by) params.set("sort_by", query.sort_by);
  if (query.limit) params.set("limit", String(query.limit));
  const qs = params.toString();
  return fetchApi<Neighborhood[]>(`/neighborhoods${qs ? `?${qs}` : ""}`);
}

export async function getFeaturedNeighborhoods(limit: number = 8): Promise<FeaturedNeighborhoodsResponse> {
  return fetchApi<FeaturedNeighborhoodsResponse>(`/neighborhoods/featured?limit=${limit}`);
}

File: backend/app/api/v1/neighborhoods.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List, Dict, Any
from backend.app.db.session import get_db
from backend.app.services.neighborhood_service import get_neighborhoods, get_featured_neighborhoods

router = APIRouter()

@router.get("/", response_model=List[Dict[str, Any]])
async def read_neighborhoods(
    sort_by: str = Query("median_sold_price", pattern="^(median_sold_price|average_sold_price|avg_asking_price|locality)$"),
    limit: int = Query(50, le=200),
    db: AsyncSession = Depends(get_db),
):
    return await get_neighborhoods(db, sort_by=sort_by, limit=limit)

@router.get("/featured", response_model=Dict[str, Any])
async def read_featured_neighborhoods(
    limit: int = Query(8, le=20),
    db: AsyncSession = Depends(get_db),
):
    return await get_featured_neighborhoods(db, limit=limit)
